web/components/roomConfig/GitlabRepoConfig.tsx

```diff
@@ -0,0 +1,167 @@
+import { h, FunctionComponent, createRef } from "preact";
+import { useState, useCallback, useEffect } from "preact/hooks";
+import BridgeAPI from "../../BridgeAPI";
+import { ConnectionConfigurationProps, RoomConfig } from "./RoomConfig";
+import { GitLabRepoConnectionState, GitLabRepoResponseItem, GitLabTargetFilter, GitLabRepoConnectionTarget, GitLabRepoConnectionProjectTarget, GitLabRepoConnectionInstanceTarget } from "../../../src/Connections/GitlabRepo";
+import InputField from "../InputField";
+import ButtonSet from "../ButtonSet";
+import { Button } from "../Button";
+
+const EventType = "uk.half-shot.matrix-hookshot.gitlab.repository";
+
+const ConnectionSearch: FunctionComponent<{api: BridgeAPI, onPicked: (state: GitLabRepoConnectionState) => void}> = ({api, onPicked}) => {
+    const [filter, setFilter] = useState<GitLabTargetFilter>({});
+    const [results, setResults] = useState<GitLabRepoConnectionProjectTarget[]|null>(null);
+    const [instances, setInstances] = useState<GitLabRepoConnectionInstanceTarget[]|null>(null);
+    const [debounceTimer, setDebounceTimer] = useState<number>(null);
+    const [currentProjectPath, setCurrentProjectPath] = useState<string|null>(null);
+
+    const searchFn = useCallback(() => {
+        api.getConnectionTargets<GitLabRepoConnectionTarget>(EventType, filter).then(res => {
+            if (!filter.instance) {
+                setInstances(res);
+                if (res[0]) {
+                    setFilter({instance: res[0].name});
+                }
+            } else {
+                setResults(res as GitLabRepoConnectionProjectTarget[]);
+            }
+        });
+    }, [api, filter]);
+
+    const updateSearchFn = useCallback((evt: InputEvent) => {
+        setFilter({...filter, search: (evt.target as HTMLInputElement).value });
+    }, [filter, setFilter]);
```

Comment:
Setters from `useState` never change.
```suggestion
    }, [filter]);
```